src/apps/common/bip32.c

```diff
@@ -16,6 +16,8 @@
 #include <hardfault.h>
 #include <util.h>
 
+#include <wally_bip32.h>
```

Comment:
It is used there, `struct ext_key` comes from it. I'll remove it here then as it is already included.

Reply:
ah! `ext_key`. very sneaky :)

> I'll remove it here then as it is already included.

I'm used to a practice to always include headers if any definition is used in the code. So, I'd keep it here. It was a good catch. I just missed `ext_key` in bip32.h. I should've never made the original comment :laughing:

Reply:
> I'm used to a practice to always include headers if any definition is used in the code

I do to, with one exception: stuff included in the .h file corresponding to the .c file. Otherwise you have to maintain a 1:1 copy of all includes of the header again in the .c file. I could live with it, but also not worth spending time on it, as most of the codebase adheres to this (or tries to ;)).

Reply:
> Otherwise you have to maintain a 1:1 copy of all includes of the header again in the .c file

This is the way.

![image](https://user-images.githubusercontent.com/25405/100628976-1261b100-3329-11eb-9e73-5181af4de289.png)

Reply:
I know, but practically speaking, is it worth it? When is that actually relevant? As far as I can tell, only when you copy paste snippets of code, but then it is also no big deal to go to the header and copy the relevant includes as well. What other issues can happen?

Reply:
It's inconsistent. bip32.c now includes *some* headers but not the others which are in bip32.h. Things move around, generate unnecessary changes in diff. Headers are for other modules to use, .c files are implementations. You know all this :)

Suppose some functions in bip32.c become a specific to the implementation and `ext_key` is not visible anywhere in bip32.h. Now you'll need to move wally_bip32.h include from bip32.h back to bip32.c. Otherwise, the "includers" of bip32 also get wally_bip32.h for no reason.

These small inconsistencies eventually add up and lead to all sorts of problems: slower compilation times, harder to refactor, and maybe even include conflicts.

Reply:
> bip32.c now includes some headers but not the others which are in bip32.h. 

I checked, that is not true. No include that is in bip32.h is also in bip32.c currently.

>  Otherwise, the "includers" of bip32 also get wally_bip32.h for no reason.

That is true regardless of whether you had the include in the .c file or not. 

> These small inconsistencies eventually add up and lead to all sorts of problems: slower compilation times, harder to refactor, and maybe even include conflicts.

I am thinking about it but I can't find an example where not having the include in the .c file can lead to any of those problems.

I do know that many (or most?) say that all includes should be in the .c file, and on the surface it makes sense; I just never saw a compelling reason to copy/paste (and keep both copies in sync) includes that are already present in the .h file belonging to the .c file.

Reply:
I just gave you a compelling reason above:

> These small inconsistencies eventually add up and lead to all sorts of problems: slower compilation times, harder to refactor, and maybe even include conflicts.

The bip32 case is just one example. You'll want to apply the same consistent method to the whole codebase. If you do it with the reasoning where headers are scattered around, you'll get a very compelling reason in the end :)
This doesn't apply specifically to just bip32.h and bip32.c.

Reply:
> > Otherwise, the "includers" of bip32 also get wally_bip32.h for no reason.
> 
> That is true regardless of whether you had the include in the .c file or not.

here's a hypothetical bad example:

- bip32.h <- includes wally.h because of some "key" type
- bip32.c <- includes bip32.h
- app.c <- includes bip32.h and wally.h to call a func in bip32

now, suppose bip32.c wants to change and abstract the types or change implementation:

- bip32.h <- still includes wally.h but need not to
- bip32.c <- includes bip32.h
- app.c <- includes bip32.h; it also gets wally.h although it doesn't need nor wants it

if say app happens to use another libmolly which has the same "key" type for other reasons, it may very well be that the types from wally.h and molly.h will conflict due to app.c is fed wally.h against its will.

the standard way of doing it is:

- bip32.h <- doesn't include wally.h because it uses no wally's types in declarations
- bip32.c <- includes bip32.h and wally.h because it uses libwally underneath
- app.c <- includes bip32.h and doesn't get wally.h, as it should

when the app binary is produced, app is linked against bip32.o. at no point does app.c get wally.h included in the source.

again, this is a very small example. if you extrapolate it to a relatively big codebase, you'll get bigger problems, more tech. debt, compilation times, conflicts, etc. 

besides, I remember you once said you like consistency :)